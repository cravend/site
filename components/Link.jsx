import React from 'react';
import NextLink from 'next/link';
import { Link as StyledLink } from 'theme-ui';
import { string, node } from 'prop-types';

const Link = ({ to, children, ...props }) => {
  const externalCheck = /^https?:\/\//;
  if (externalCheck.test(to)) {
    return <StyledLink href={to} {...props}>{children}</StyledLink>;
  }
  return <NextLink passHref href={to} {...props}><StyledLink>{children}</StyledLink></NextLink>;
};

Link.propTypes = {
  to: string.isRequired,
  children: node.isRequired,
};

export default Link;
