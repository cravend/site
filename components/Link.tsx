import React, { FC, ReactNode } from "react";

import NextLink from "next/link";
import { Link as StyledLink } from "theme-ui";

import { Locale } from "../src/translations/types";

type LinkType = {
  to: string;
  isExternal?: boolean;
  locale?: Locale;
  color?: string | string[];
  children: ReactNode;
};

const Link: FC<LinkType> = ({
  to,
  isExternal,
  locale,
  color,
  children,
}: LinkType) => {
  const externalCheck = /^https?:\/\//;
  if (isExternal || externalCheck.test(to)) {
    return (
      <StyledLink sx={{ color }} href={to}>
        {children}
      </StyledLink>
    );
  }
  return (
    <NextLink passHref locale={locale} href={to}>
      <StyledLink sx={{ color }}>{children}</StyledLink>
    </NextLink>
  );
};

Link.defaultProps = {
  isExternal: false,
  locale: undefined,
  color: undefined,
};

export default Link;
