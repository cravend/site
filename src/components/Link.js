import React from 'react'
import styled from '@emotion/styled'
import { Link as GatsbyLink } from 'gatsby'
import { OutboundLink } from 'gatsby-plugin-google-analytics'

export const Link = ({
  children,
  to,
  activeClassName,
  partiallyActive,
  ...other
}) => {
  const internal = /^\/(?!\/)/.test(to)
  const file = /\.[0-9a-z]+$/i.test(to)

  if (internal) {
    if (file) {
      return (
        <OutboundLink href={to} {...other}>
          {children}
        </OutboundLink>
      )
    }
    return (
      <GatsbyLink
        to={to}
        activeClassName={activeClassName}
        partiallyActive={partiallyActive}
        {...other}
      >
        {children}
      </GatsbyLink>
    )
  }
  return (
    <a href={to} {...other}>
      {children}
    </a>
  )
}

export const StyledLink = styled(Link)`
  text-decoration: none;
  font-weight: 600;
  color: inherit;
  border-bottom: 2px solid;
  :hover,
  :focus,
  :active {
    border-bottom: 3px solid #a8dcd9;
  }
`

export const StyledLinkOld = styled(Link)`
  text-decoration: none;
  font-weight: 600;
  color: inherit;
  position: relative;
  overflow: hidden;
  :before {
    content: '';
    position: absolute;
    left: 0;
    right: 0;
    bottom: -1px;
    background-color: #fff;
    height: 2px;
    transition-property: background-color;
    transition-duration: 0.3s;
    transition-timing-function: ease-out;
  }
  :hover:before,
  :focus:before,
  :active:before {
    background-color: #a8dcd9;
    left: 0;
    right: 0;
  }
`
export default Link
