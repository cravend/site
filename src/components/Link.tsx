import NextLink from "next/link";
import styles from "../styles/modules/Link.module.scss";
import type { Locale } from "../i18n/config";
import type { FC, ReactNode } from "react";

type LinkType = {
  to: string;
  isExternal?: boolean;
  locale?: Locale;
  children: ReactNode;
};

const Link: FC<LinkType> = ({ to, isExternal, locale, children }: LinkType) => {
  const externalCheck = /^https?:\/\//;
  if (isExternal || externalCheck.test(to)) {
    return (
      <a className={styles.primary} href={to}>
        {children}
      </a>
    );
  }
  return (
    <NextLink locale={locale} href={to}>
      {/* eslint-disable-next-line jsx-a11y/anchor-is-valid */}
      <a className={styles.primary}>{children}</a>
    </NextLink>
  );
};

export default Link;
