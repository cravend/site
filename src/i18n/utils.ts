import { locales, defaultLocale, RTL_LOCALES } from "./config";

import type { Locale } from "./config";
import type { NextRouter } from "next/router";

/**
 * @description This function is used to check if a string is a valid locale.
 *
 * @param localeToTest The string to check.
 * @returns True if the string is a valid locale, false otherwise.
 */
export const isLocale = (localeToTest?: string): localeToTest is Locale =>
  locales.some((locale) => locale === localeToTest);

/**
 * @description This function is used to get the current locale.
 *
 * If a custom locale is set by \@next/router, it will be returned.
 * If no custom locale is set, the router's default locale will be returned.
 * If the router's default locale isn't set, the default locale will be returned.
 *
 * @param router The NextRouter instance.
 * @returns The current locale.
 */
export const getLocale = (router: NextRouter | null) => {
  if (!router) return defaultLocale;
  if (isLocale(router.locale)) return router.locale;
  if (isLocale(router.defaultLocale)) return router.defaultLocale;
  return defaultLocale;
};

/**
 * @description This function is used to toggle the current locale.
 *
 * @param locale The current locale to be toggled.
 * @returns The new locale.
 */
export const toggleLocale = (locale: Locale) =>
  locale === locales[0] ? locales[1] : locales[0];

/**
 * @description This function is used to get the text direction of the current locale.
 *
 * @param locale The locale to be checked
 * @returns True if the locale is read right-to-left, false otherwise
 */
export const isRtl = (locale: Locale) => RTL_LOCALES.includes(locale);
