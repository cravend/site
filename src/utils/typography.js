import Typography from 'typography'

const typography = new Typography({
  baseLineHeight: 1.666,
  headerFontFamily: ['Poppins', 'Helvetica', 'Arial', 'sans-serif'],
  bodyFontFamily: ['Poppins', 'Helvetica', 'Arial', 'sans-serif'],
  googleFonts: [
    {
      name: 'Poppins',
      styles: ['300', '600', '700']
    }
  ]
})

export default typography
