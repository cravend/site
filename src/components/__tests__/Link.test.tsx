import { fireEvent, screen, render } from "../../tests/utils";
import Link from "../Link";

describe("<Link />", () => {
  it("renders", () => {
    expect.assertions(1);
    const LINK_TEXT = "Hello, world!";
    const { container } = render(<Link to="/">{LINK_TEXT}</Link>);
    expect(container).toMatchSnapshot();
  });

  it("pushes to another page when clicked", () => {
    expect.hasAssertions();
    const LINK_TEXT = "Hello, world!";
    const pushMock = jest.fn();

    render(<Link to="/test">{LINK_TEXT}</Link>, {
      router: { push: pushMock },
    });
    fireEvent.click(screen.getByText(LINK_TEXT));

    expect(pushMock).toHaveBeenCalledTimes(1);
    expect(pushMock).toHaveBeenLastCalledWith("/test", "/test", {});
  });

  it("renders an anchor tag for external HTTP links", () => {
    expect.assertions(2);
    const LINK_TEXT = "Hello, world!";
    const pushMock = jest.fn();

    render(<Link to="http://www.test.com/">{LINK_TEXT}</Link>, {
      router: { push: pushMock },
    });
    fireEvent.click(screen.getByText(LINK_TEXT));

    expect(pushMock).not.toHaveBeenCalled();
    expect(screen.getByRole("link", { name: LINK_TEXT })).toHaveAttribute(
      "href",
      "http://www.test.com/"
    );
  });

  it("renders an anchor tag for external HTTPS links", () => {
    expect.assertions(2);
    const LINK_TEXT = "Hello, world!";
    const pushMock = jest.fn();

    render(<Link to="https://www.test.com/">{LINK_TEXT}</Link>, {
      router: { push: pushMock },
    });
    fireEvent.click(screen.getByText(LINK_TEXT));

    expect(pushMock).not.toHaveBeenCalled();
    expect(screen.getByRole("link", { name: LINK_TEXT })).toHaveAttribute(
      "href",
      "https://www.test.com/"
    );
  });

  it("renders an anchor tag for forced external links", () => {
    expect.assertions(2);
    const LINK_TEXT = "Hello, world!";
    const pushMock = jest.fn();

    render(
      <Link to="/" isExternal>
        {LINK_TEXT}
      </Link>,
      { router: { push: pushMock } }
    );
    fireEvent.click(screen.getByText(LINK_TEXT));

    expect(pushMock).not.toHaveBeenCalled();
    expect(screen.getByRole("link", { name: LINK_TEXT })).toHaveAttribute(
      "href",
      "/"
    );
  });

  it("updates the router locale when clicked", () => {
    expect.assertions(2);
    const LINK_TEXT = "Hello, world!";
    const NEW_LOCALE = "fr";
    const pushMock = jest.fn();
    render(
      <Link to="/" locale={NEW_LOCALE}>
        {LINK_TEXT}
      </Link>,
      { router: { push: pushMock } }
    );

    fireEvent.click(screen.getByRole("link", { name: LINK_TEXT }));
    expect(pushMock).toHaveBeenCalledTimes(1);
    expect(pushMock).toHaveBeenLastCalledWith("/", "/", {
      locale: NEW_LOCALE,
      scroll: undefined,
      shallow: undefined,
    });
  });
});
