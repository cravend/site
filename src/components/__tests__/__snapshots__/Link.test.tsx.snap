// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> renders 1`] = `
<div>
  <a
    href="/"
  >
    Hello, world!
  </a>
</div>
`;
